pkg/userdata/plugin/plugin.go

```diff
@@ -0,0 +1,123 @@
+//
+// Core UserData plugin.
+//
+
+// Package plugin provides the plugin side of the plugin mechanism.
+// Individual plugins have to implement the provider interface,
+// pass it to a new plugin instance, and call run.
+package plugin
+
+import (
+	"encoding/json"
+	"fmt"
+	"net"
+	"os"
+
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
+	clusterv1alpha1 "sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
+)
+
+// Provider defines the interface each plugin has to implement
+// for the retrieval of the userdata based on the given arguments.
+type Provider interface {
+	UserData(
+		spec clusterv1alpha1.MachineSpec,
+		kubeconfig *clientcmdapi.Config,
+		cloudConfig, cloudProviderName string,
+		clusterDNSIPs []net.IP,
+		externalCloudProvider bool,
+	) (string, error)
+}
+
+// Handler cares dispatching of the RPC calls to the given Provider.
+type Handler struct {
+	provider Provider
+}
+
+// UserData receives the UserData RPC message and calls the provider.
+func (h *Handler) UserData(req *UserDataRequest, resp *UserDataResponse) error {
```

Comment:
This is unused and I am a bit confused as to why the linter doesn't complain about that :thinking:

Reply:
That's why PR is WiP. This type as well as many of the log usages and overly detailed scriptings have just been for debugging or testing. This handler here has i.e. needed for the `net.RPC` solution and will now be removed. The linter doesn't complain because it's exported and may be used by any importer in future. The linter normally only complains private unused types, functions, and variables.